Add "Drag And Drop Across Frames" keyword for cross-frame drag and drop support

b-vamsipunnam opened the pull request:
Summary

This PR adds a new keyword `Drag And Drop Across Frames` to support drag-and-drop interactions where the source and target are in different frame contexts.

Motivation

While automating post-deployment flows in a Salesforce Service Cloud setup, I encountered a limitation where drag-and-drop failed when the drop target was inside an iframe. Existing keywords do not support this scenario because WebDriver action chains lose context when switching frames mid-action.

Real-world scenario (Salesforce Service Cloud):  
Login → Setup → All Sites → Builder → Select Dashboard → Drag and drop a component into a target section (inside an iframe) → Publish

The existing `Drag And Drop` keyword fails in this case because the target resides inside an iframe.

This aligns with known limitations:
- Issue #120: cross-frame drag-and-drop challenges  
- Related discussion #1813 around iframe handling

This keyword enables reliable end-to-end automation for such scenarios.

Changes

- Added implementation in src/SeleniumLibrary/keywords/element.py
- Added acceptance tests in atest/acceptance/keywords/draganddropframe.robot
- Added test HTML page atest/resources/html/frames/draganddrop.html
- Updated unit test keyword count in utest/test/api/test_plugins.py

All acceptance tests pass locally, including the new cross-frame scenarios.

All acceptance tests pass locally (including new ones).

This addresses a long-standing limitation around cross-frame drag-and-drop. If applicable, please consider adding the ``acknowledgment`` label to #120 for release notes. Thank you!

b-vamsipunnam committed:
Add Drag And Drop To Frame keyword Updated implementation and tests

b-vamsipunnam committed:
Fix acceptance tests for CI run -  Drag And Drop To Frame keyword

yuriverweij commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -1293,4 +1293,41 @@ def get_css_property_value(
+            target_element = self.find_element(target)
+
+            action = ActionChains(self.driver, duration=self.ctx.action_chain_delay)
+            action.move_to_element(target_element).release().perform()
```
If the `find_element()` for frame or target raises an error, the mouse button is still held down. The finally switches context back but never releases. The next action in the suite inherits a held mouse button.

b-vamsipunnam replied:
@yuriverweij : I’ve updated the keyword to ensure the mouse is always released if anything fails after the click-and-hold. The `finally` block now handles the release.

yuriverweij commented on atest/acceptance/keywords/draganddropframe.robot:
```diff
@@ -0,0 +1,47 @@
+Standard Drag And Drop Fails When Target Is Inside Frame
+    [Documentation]    Verifies that the standard Drag And Drop keyword cannot complete this cross-frame scenario.
+    Wait Until Page Contains Element    id=source    timeout=10s
+    Run Keyword And Expect Error    *    Drag And Drop    id=source    id=target
```
I would not wildcard the error that is expected. This can cause the test to "pass" when any error occurs.

b-vamsipunnam replied:
@yuriverweij : I’ve updated the tests to assert the specific expected error instead of using a wildcard. Thank you.

yuriverweij commented on atest/acceptance/keywords/draganddropframe.robot:
```diff
@@ -0,0 +1,47 @@
+Drag And Drop To Frame Fails With Invalid Frame
+    [Documentation]    Verifies that the keyword fails when the frame locator is invalid.
+    Wait Until Page Contains Element    id=source    timeout=10s
+    Run Keyword And Expect Error    *    Drag And Drop To Frame
```
I would not wildcard the error that is expected. This can cause the test to "pass" when any error occurs. 
*Element with locator 'id=missingFrame' not found*

b-vamsipunnam replied:
@yuriverweij : @yuriverweij : I’ve updated the tests to assert the specific expected error instead of using a wildcard. Thank you.

yuriverweij commented on atest/acceptance/keywords/draganddropframe.robot:
```diff
@@ -0,0 +1,47 @@
+Drag And Drop To Frame Fails With Invalid Target
+    [Documentation]    Verifies that the keyword fails when the target element is not found inside the iframe.
+    Wait Until Page Contains Element    id=source    timeout=10s
+    Run Keyword And Expect Error    *    Drag And Drop To Frame
```
I would not wildcard the error that is expected. This can cause the test to "pass" when any error occurs.
*Element with locator 'id=missingTarget' not found*

b-vamsipunnam replied:
@yuriverweij : I’ve updated the tests to assert the specific expected error instead of using a wildcard. Thank you.

yuriverweij commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -1293,4 +1293,41 @@ def get_css_property_value(
+
+    @keyword('Drag And Drop To Frame')
+    def drag_and_drop_to_frame(
+            self, locator: Locator, target: Locator, frame: Locator,
```
In order to make the keyword more generic it might be worth changing some parts:

`def drag_and_drop_across_frames(
        self, locator: Locator, 
        target: Locator,
        target_frame: Locator, 
        source_frame: Optional[Locator] = None,
`

- ``locator``: Locator of the source element to drag.
- ``target``: Locator of the drop target.
- ``target_frame``: Locator of the iframe containing the target.
- ``source_frame``: Locator of the iframe containing the source, or None if the source is in the default content.

b-vamsipunnam replied:
@yuriverweij : I’ve updated the keyword signature and documentation to make it more generic:`locator`, `target`, `target_frame`, and optional `source_frame`.This now supports both default-content to frame and source-frame to target-frame scenarios. Thanks for the suggestion. This has now become a robust keyword.

yuriverweij commented on atest/resources/html/frames/draganddrop.html:
Test fixture does not seem to work as intended.
Right now:
When clicking on the "Drag Me!" box, and performing a mouseup (releasing the mouse button anywhere) will trigger "Dropped Successfully", regardless of actually moving/draggint the box.

The test would pass even if move_to_element(target_element) did nothing.
The test would pass even if the frame switch silently failed (as long as a mouseup fires somewhere).
The test would pass if you replaced the whole keyword body with click_and_hold().release() on the source.

The fixture needs to validate that the drop actually lands on the target inside the frame,  e.g. using real HTML5 DnD events or at least checking the mouseup coordinates hit the target's bounding box. As written, these tests give false confidence

b-vamsipunnam replied:
Hi @yuriverweij, thanks for the detailed feedback. I agree with the suggestions and I’m working on the updates. I’ll push the changes shortly.

b-vamsipunnam replied:
@yuriverweij : I updated the fixture so it only completes the drop when the mouseup coordinates are inside the target element within the target iframe. This should prevent false positives from a simple click-and-release on the source. Thanks you.

yuriverweij reviewed (changes requested):
Thanks for this PR! it looks very promising, but there are a few things that need to be resolved.

- There is a bug in the try/except which can cause the mouse button to stay down.
- The test fixture (html file) has some problems which can give false positives 
- Suggestion: the keyword can be made a bit more generic allowing dragging from frame to frame or default context to frame. (in this case the Note can be removed from the docstring)
- Suggestion: add the following to the docstring/description `After this keyword runs, the browser context is always default content, regardless of the context before the call.`

b-vamsipunnam committed:
Improve cross-frame drag and drop reliability and tests

b-vamsipunnam commented:
Hi @yuriverweij

Thanks for the detailed feedback.

I’ve updated the fixture so the drop completes only when the mouseup occurs within the target element inside the target iframe, avoiding false positives.

The keyword now also supports an optional source frame and ensures the mouse is released on failure before resetting context.

@emanlove: If this looks good, could you please consider adding the ``acknowledgment`` label to #120 for release notes? Thank you.

b-vamsipunnam committed:
Update docstring to clarify default content reset after execution

yuriverweij commented on atest/resources/html/frames/draganddrop.html:
```diff
@@ -2,94 +2,262 @@
+</div>
+
+<div id="dragGhost"></div>
+<div id="status">Not dropped</div>
```
Nitpicking/Suggestion:
`#status` isn't used in the tests, instead they assert on the `id=target` (both are updated on successful drop) 

So you could remove the `#status` entirely, or use it in the tests. 
I think it is best to keep the test fixture as small and clean as possible.

b-vamsipunnam replied:
@yuriverweij : As suggested, I removed the unused #status.

yuriverweij commented on atest/acceptance/keywords/draganddropframe.robot:
```diff
@@ -1,47 +1,87 @@
+    Element Should Contain    id=target    Dropped Successfully!
+    Unselect Frame
+
+Drag And Drop Across Frames Returns To Default Content
```
This test doesn't really verify what it claims. 

`Element Should Be Visible    id=targetFrame` checks that the iframe element is visible, but the iframe element lives in the default content regardless of which frame context Selenium currently has selected. 
If the keyword left us stuck inside the target frame, this assertion would either still pass (because the lookup falls through) or fail for an unrelated reason. Neither outcome proves the context was restored.

Suggested fix: assert on something that's not reachable when we're actually back in default content:
`Page Should Not Contain Element    id=target`
`id=target` only exists inside `targetFrame`. If the keyword correctly returned to default content, the test passes.
It would also catch a bug in case we're accidentally in some other non-default frame.

b-vamsipunnam replied:
@yuriverweij : You are right.  I have updated the test to assert on an element only available inside the iframe, which validates that the context is correctly reset to default content.

yuriverweij commented on atest/acceptance/keywords/draganddropframe.robot:
```diff
@@ -1,47 +1,87 @@
+    Element Should Not Be Visible    id=frameSource
+    Unselect Frame
 
 Standard Drag And Drop Fails When Target Is Inside Frame
```
I'd suggest dropping this test. It's not testing the new Drag And Drop Across Frames keyword, it's testing the standard `Drag And Drop` keyword limitations.

b-vamsipunnam replied:
@yuriverweij : as suggested, I have removed this test. Thank you.

yuriverweij reviewed (changes requested):
On test doesn't do what it claims to do. For the rest it looks good to me. 

Maybe @emanlove still has some other suggestions/feedback and I'll let him have the final say.

b-vamsipunnam committed:
Simplify test fixture and refine test coverage

b-vamsipunnam commented:
Hi @yuriverweij : Thanks for all the feedback. I’ve addressed the comments and made the suggested updates. Happy to make any further changes if needed.

Thank you.

yuriverweij reviewed (approved):
Looks good to me!

emanlove committed:
Merge branch 'master' into feature/drag-and-drop-to-frame
